workspaces/liferay-dlfoldertemplate-workspace/client-extensions/liferay-dlfoldertemplate-custom-element/src/components/template-diagram/Diagram.js

```diff
@@ -0,0 +1,505 @@
+/**
+ * SPDX-FileCopyrightText: (c) 2000 Liferay, Inc. https://liferay.com
+ * SPDX-License-Identifier: LGPL-2.1-or-later OR LicenseRef-Liferay-DXP-EULA-2.0.0-2023-06
+ */
+
+import dagre from 'dagre';
+import React, {useCallback, useEffect, useState} from 'react';
+import ReactFlow, {
+	Background,
+	ConnectionLineType,
+	Controls,
+	Panel,
+	useEdgesState,
+	useNodesState,
+} from 'reactflow';
+
+import 'reactflow/dist/style.css';
+import ClayButton, {ClayButtonWithIcon} from '@clayui/button';
+import {ClayInput} from '@clayui/form';
+import {Form, Input} from 'antd';
+
+import {
+	addNode,
+	deleteFolderTemplateBatch,
+	getAvailableTemplatesNodesPage,
+	updateFolderTemplate,
+} from '../../services/TemplateDiagramService';
+import {showError} from '../../utils/util';
+import FolderNode from './controls/custom-node/FolderNode';
+
+import './Diagram.css';
+
+const dagreGraph = new dagre.graphlib.Graph();
+
+dagreGraph.setDefaultEdgeLabel(() => ({}));
+
+const nodeWidth = 172;
+
+const nodeHeight = 36;
+
+const position = {x: 0, y: 0};
+
+const edgeType = 'SmoothStep';
+
+const getLayoutedElements = (nodes, edges, direction = 'TB') => {
+	dagreGraph.setGraph({
+		rankdir: direction,
+	});
+
+	nodes.forEach((node) => {
+		dagreGraph.setNode(node.id, {height: nodeHeight, width: nodeWidth});
```

Comment:
if i got your comment in the right way, i need this to be triggered each time i add new node, since i need to recalculate the new and the current node positions, so the dagreGraph.setNode will do the required